src/lowering/desugar.ml

```diff
@@ -239,6 +242,47 @@ and lexp' = function
   | S.IdxE (e1, e2) -> I.IdxLE (exp e1, exp e2)
   | _ -> raise (Invalid_argument ("Unexpected expression as lvalue"))
 
+and rewrite_for_to_while p arr proj c0 c1 c2 e1 e2 =
+  (* for (p in e1) e2 when e1 = ⟨...array-typed...⟩.vals()
+     ~~>
+     let arr = ⟨...array-typed...⟩ ;
+     let size = arr.size() ;
+     var indx = 0 ;
+     label l loop {
+       if indx < size
+       then { let p = arr[indx]; e2; indx += 1 }
+       else { break l }
+     } *)
+  let arr_typ = arr.note.note_typ in
+  let arrv = fresh_var "arr" arr_typ in
+  let size_exp =
+    let arr_bind =
+      { arr with it = VarE { it = id_of_var arrv; note = (); at = arr.at };
+                 note = { note_typ = arr.note.note_typ; note_eff = T.Triv } } in
+    exp { e1 with note = { note_typ = T.nat; note_eff = c2.note.note_eff };
+                  it = CallE ({ note = { note_typ = T.(Func (Local, Returns, [], [], [nat]));
+                                         note_eff = c2.note.note_eff };
+                                it = DotE (arr_bind, { proj with it = "size" });
+                                at = c0 },
+                              c1, c2) } in
```

Comment:
Ok, I was wrong, and missed the fact that you are invoking `exp` on source.
But I still think that's a little indirect and hard to follow.
I would delete `arr_bind` and just insert  a `expD (exp c2)` and the simpler size_exp suggested by @nomeata
(see below)